### PR TITLE
when-clause parser improvements/fixes

### PR DESCRIPTION
**Pull request by ulugbekna**
- context keys: remove unused new parser caller (we don't really need it right now, especially given the parser API keeps changing)
- context keys: parser: support for running without regex parsing error recovery
- context keys: parsing: rewrite scanner and parser error reporting
- context keys: parsing: raise parsing error on regexp object creation
- context keys: parser: preserve old parsers behavior with "foo == 'true'" remaining same, but "foo == true" being optimized to "foo"

### CODE REVIEW

File: src/vs/platform/contextkey/test/common/parser.test.ts [hunk 1]
```diff
@@ -173,7 +172,7 @@ suite('Context Key Parser', () => {
 
 		test(`/foo`, () => {
 			const input = `/foo`;
-			assert.deepStrictEqual(parseToStr(input), "Lexing errors:\n\nUnexpected token '/foo' at offset 0\n\n --- \nParsing errors:\n\nExpected 'true', 'false', '(', KEY, KEY '=~' regex, KEY [ ('==' | '!=' | '<' | '<=' | '>' | '>=' | 'in' | 'not' 'in') value ] but got '/foo' at offset 0.\n");
+			assert.deepStrictEqual(parseToStr(input), "Lexing errors:\n\nUnexpected token '/foo' at offset 0. Did you forget to escape the '/' (slash) character? Put two backslashes before it to escape, e.g., '\\\\/'.\n\n --- \nParsing errors:\n\nUnexpected '/foo' at offset 0.\n");
```
Comment (jrieken) [inline src/vs/platform/contextkey/test/common/parser.test.ts:175]: This is totally fine but might be "future-fragile". Meaning that you'll encounter a test failure whenever the string/message is tweaked. You could consider a static error code in addition to the messages and use them here

**Review by jrieken — DISMISSED** [review]
Looking very good. Left a nit but this good to merged.

File: src/vs/platform/contextkey/test/common/parser.test.ts (hunk 1 above)
Comment (ulugbekna) [reply src/vs/platform/contextkey/test/common/parser.test.ts:175]: Oh, I had to suffer the pain of fragility, but I would like to go about this problem by fixing https://github.com/microsoft/vscode/issues/174251

File: src/vs/platform/contextkey/common/contextkey.ts [hunk 2]
```diff
@@ -321,16 +333,19 @@ export class Parser {
...
+	get lexingErrors(): Readonly<LexingError[]> {
+		return this._scanner.errors;
 	}
 
-	get parsingErrors(): Readonly<string[]> {
+	get parsingErrors(): Readonly<ParsingError[]> {
 		return this._parsingErrors;
 	}
 
+	constructor(private _config: ParserConfig = defaultConfig) {
```
[the remark this answers is not in the document]
Comment (ulugbekna) [reply src/vs/platform/contextkey/common/contextkey.ts]: good catch, thanks! fixed and pushed

**Review by jrieken — APPROVED** [review]

**Pull request merged**